capreolus/trainer/__init__.py

```diff
@@ -292,3 +322,265 @@ def predict(self, reranker, pred_data, pred_fn):
+            for sample in dataset
+        ]
+
+        return [self.write_tf_record_to_file(dir_name, tf_features)]
+
+    def convert_to_tf_train_record(self, dataset):
+        """
+        Tensorflow works better if the input data is fed in as tfrecords
+        Takes in a dataset,  iterates through it, and creates multiple tf records from it.
+        """
+        tf_record_filenames = []
+        tf_features = []
+        dir_name = self.get_cache_path() / dataset.get_hash()
+
+        # There are 'n' iterations
+        # Each iterations has a size 'itersize' number of batches in it
+        for niter in tqdm(range(0, self.cfg["niters"]), desc="Converting data to tf records"):
+            for sample_idx, data in enumerate(dataset):
+                # TODO: Split into multiple files? This might be too much memory consumption
+                tf_features.append(
+                    self.create_tf_feature(
+                        data["qid"], data["query"], data["query_idf"], data["posdocid"], data["posdoc"],
+                        data["negdocid"], data["negdoc"]
+                    )
+                )
+
+                # TODO: Below line means that the tfrecord created varies based on itersize and batch. Might interfere with caching
+                if sample_idx + 1 >= self.cfg["itersize"] * self.cfg["batch"]:
+                    break
+
+            if (niter + 1) % 10 == 0:
+                tf_record_filenames.append(self.write_tf_record_to_file(dir_name, tf_features))
+                tf_features = []
+
+        if len(tf_features):
+            tf_record_filenames.append(
+                self.write_tf_record_to_file(dir_name, tf_features)
+            )
+
+        return tf_record_filenames
+
+    def cache_exists(self, dataset):
+        # TODO: The caching logic is broken - the cache cannot be reused if itersize/batch size e.t.c changes
+        cache_dir_name = dataset.get_hash()
+        cache_dir_path = self.get_cache_path() / cache_dir_name
+        # TODO: Add checks to make sure that the number of files in the director is correct
+        return os.path.isdir(cache_dir_path) and len(os.listdir(cache_dir_path)) != 0
+
+    def load_tf_records_from_file(self, filenames):
+        raw_dataset = tf.data.TFRecordDataset(filenames)
+        feature_description = {
+            'qid': tf.io.FixedLenFeature([], tf.string),
+            'query': tf.io.FixedLenFeature([self.cfg["maxqlen"]], tf.float32),
+            'query_idf': tf.io.FixedLenFeature([self.cfg["maxqlen"]], tf.float32),
+            'posdoc_id': tf.io.FixedLenFeature([], tf.string),
+            'posdoc': tf.io.FixedLenFeature([self.cfg["maxdoclen"]], tf.float32),
+            'negdoc_id': tf.io.FixedLenFeature([], tf.string, default_value=b'na'),
+            'negdoc': tf.io.FixedLenFeature([self.cfg['maxdoclen']], tf.float32, default_value=tf.zeros(self.cfg['maxdoclen']))
```

Comment:
Will the cache be broken if `maxqlen` or `maxdoclen` changes?

Reply:
Hmmm yes it would be. We iterate through the dataset, convert everything using the `extractor` and then cache the results as tf records. If the maxqlen changes, or if the embedding (glove6b) changes, then the cache would need to be invalidated.

I'll add the extractor config as well contribute to the dataset's `__hash()`